### PR TITLE
[Feature] Grading statistics

### PR DESCRIPTION
**Pull request by schultek**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Add statistics to the grading page for instructors.

##### Charts & Statistics
- Added a Server REST call to generate the statistics
- Added charts for each test case showing the percentage of students that passed the test
- Added charts for each sca category showing the amount of issues detected across students
- Added charts area next to the datatable
  - Created a new chart component as a wrapper around the chartjs directive
  - Added one chart for each table showing different metrics regarding the test cases / sca categories.

##### Other
- Made some implovements to the overall layout of the grading page.
- Added '?' icons for the most important columns of the tables.
- Included the colors from the charts in the datatable.

### Steps for Testing

1. Create a programming exercise
2. Particpate in the exercise with multiple accounts and different solutions
    - Solve different test cases
    - Solve / Introduce different sca issues
3. Open the grading page and check the statistics
     - Tip: All charts have tooltips
4. Edit the test case and category configuration
    - The charts should update according to the changes
6. Save your changes and trigger a rebuild
7. See the updated results as the students

##### Server
- **ProgrammingExerciseGradingService**: 93% stmts
- **ProgrammingExerciseGradingStatisticsDTR**: 86% stmts

##### Client
- **programming-exercise-configure-grading.component.ts**: 82.35% stmts
- **sca-category-distribution-chart.component.ts**: 89% stmts
- **test-case-distribution-chart.component.ts**: 100% stmts
- **test-case-passed-builds-chart.component.ts**: 100% stmts
- **category-issues-chart.component.ts**: 100% stmts

*(More elaborate tests and load tests will be added in a follow-up PR)*

### Screenshots

Bildschirmfoto 2020-10-24 um 17 57 46
Bildschirmfoto 2020-10-24 um 17 57 57

### CODE REVIEW

**Review by tiffca — APPROVED** [review]
Testing went well. 
Tested a few different scenarios (test server 1): 
User 1: all correct, (initial score) 100% 
User 2: all correct but static code analysis error 100%, 
User 3: 1 correct, 7%, 
User 4: 6/13 correct 46%, 
User 6: 6 correct but static code analysis error. 46%. 
Initial scores: 
image


After adding a penalty for static code analysis errors: 
image

Also tested changing the weights of different test cases and these updated the student scores as expected. 

The charts make sense and updated as expected, including the previews. 


Students saw the change in scores.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java [hunk 1]
```diff
@@ -456,4 +461,120 @@ private void removeAllTestCaseFeedbackAndSetScoreToZero(Result result, List<Feed
         return testCase -> result.getFeedbacks().stream().noneMatch(feedback -> feedback.getText().equals(testCase.getTestName()));
     }
 
+    /**
+     * Calculates the statistics for the grading page.
+     * @param exerciseId The current exercise
+     * @return The statistics object
+     */
+    public ProgrammingExerciseGradingStatisticsDTO generateGradingStatistics(Long exerciseId) {
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java:469]: @schultek I think Codacy is right, please write shorter methods!

**Comment by schultek** [comment]
@tiffca Thanks for testing this.

> I did notice weird behavior with the building and testing seeming to take a long time, but assuming that is outside the scope of the pr. In Firefox I had to refresh the browser to see the score sometimes.

You are right, currently the client does not track the status of the builds or get notified when new results are ready. This will be added in a later PR.

**Review by fde312 — CHANGES_REQUESTED** [review]
The progress bar is not displaying anything at the moment
image

There is probably something wrong with the calculation of the points as well:
image

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseGradingResource.java [hunk 2]
```diff
@@ -94,4 +97,28 @@ public ProgrammingExerciseGradingResource(ProgrammingExerciseGradingService prog
...
+    @GetMapping(STATISTICS)
+    @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<ProgrammingExerciseGradingStatisticsDTO> getGradingStatistics(@PathVariable Long exerciseId) {
+        log.debug("REST request to get test case statistics for programming exercise {}", exerciseId);
+        ProgrammingExercise programmingExercise = programmingExerciseService.findWithTemplateParticipationAndSolutionParticipationById(exerciseId);
+
+        Course course = programmingExercise.getCourseViaExerciseGroupOrCourseMember();
+        User user = userService.getUserWithGroupsAndAuthorities();
+
+        if (!authCheckService.isAtLeastTeachingAssistantInCourse(course, user)) {
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseGradingResource.java:116]: I guess this should be "at lest instructor" based on the `@PreAuthorize`

File: src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+})
+export class TestCaseDistributionChartComponent implements OnChanges {
+    @Input() testCases: ProgrammingExerciseTestCase[];
+    @Input() testCaseStatsMap?: TestCaseStatsMap;
+    @Input() totalParticipations?: number;
+    @Input() exercise: ProgrammingExercise;
+
+    @Output() testCaseColorsChange = new EventEmitter<{}>();
+
+    chartPreset = new HorizontalStackedBarChartPreset(['Weight', 'Weight & Bonus', 'Points'], ['all weights', 'all weights and bonuses', 'all achievable points']);
```
Comment (MaisiKoleni) [inline src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts:30]:
```suggestion
    chartPreset = new HorizontalStackedBarChartPreset(['Weight', 'Weight & Bonus', 'Points'], ['all weights', 'all weights and bonuses', 'all achieved points']);
```

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
I would also recommend disabling the animation when adjusting the values in the test case section and update it instantly. We should also describe in more detail what the charts display.

The test case chart is still missing the color bars.

Sometimes the layout is inconstient when adjusting to the windows or switching between the test case and SCA tabs.

Also see two comments in the code.

Looks good in general however.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseGradingResource.java (hunk 2 above)
Comment (schultek) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseGradingResource.java]: This is fixed [4a1c028](https://github.com/ls1intum/Artemis/pull/2172/commits/4a1c028de80261f3ae08b605aab58e458489b1e6)

**Review by sleiss — APPROVED** [review]
Tested it (no code review), works fine. Some minor issues were detected (only some client styling improvements), but they are not critical.

**Review by MaisiKoleni — DISMISSED** [review]
Looks good to me as far as I can tell now, nice addition to the grading page.

There are still some small style issues that can be improved, but those can be done in a follow up as well, I think.

**Review by MaisiKoleni — APPROVED** [review]
Still looks good to me, nice to see the test case bars display not found test / failed builds.

**Review by sleiss — APPROVED** [review]
Re-approve

**Review by fde312 — APPROVED** [review]
Worked for me as well! :)

**Pull request merged**